### PR TITLE
[Merged by Bors] - feat: continuity of the parametric set integral

### PR DESCRIPTION
**Pull request by grunweg**
From the sphere eversion project.

In passing, we rename variables in one more lemma and use fun_prop in a tiny way.

-------------

- [x] depends on: #11131
- [x] depends on: #11137
- [x] depends on: #11139
- [x] depends on: #11160

### CODE REVIEW

File: Mathlib/MeasureTheory/Integral/SetIntegral.lean [hunk 1]
```diff
@@ -1409,7 +1409,7 @@ section ParametricIntegral
 
 variable [NormedAddCommGroup E]
 
-variable {α β F G 𝕜 : Type*} [TopologicalSpace α] [TopologicalSpace β] [MeasurableSpace β]
+variable {X β F G 𝕜 : Type*} [TopologicalSpace X] [TopologicalSpace β] [MeasurableSpace β]
```
Comment (Ruben-VandeVelde) [inline Mathlib/MeasureTheory/Integral/SetIntegral.lean:1412]: Why change α but not β? (Might be better in a separate PR anyway)

**Comment by loefflerd** [comment]
Could you please extract a separate mini-PR removing the style exceptions for already-fixed overlong files? (I think two of these are my fault, sorry!)

**Comment by loefflerd** [comment]
On the topic of overlong files: we definitely don't want to add new style exceptions if we can help it. I wonder if it might make sense to have a new file called something like `DominatedConvergence.lean`, which could contain all the material currently in `Bochner.lean` that relies on dominated convergence, together with the new lemmas about set integrals from this PR. That would avoid making `SetIntegral` overlong, and would also be a step in the right direction for `Bochner.lean`.

File: Mathlib/MeasureTheory/Integral/SetIntegral.lean (hunk 1 above)
Comment (grunweg) [reply Mathlib/MeasureTheory/Integral/SetIntegral.lean:1412]: Because `X` occurs in the following lemma :-)
I've renamed all `α` in this file, filed as #11131.

**Comment by grunweg** [comment]
@loefflerd Filed #11137 for the style exceptions.

**Comment by grunweg** [comment]
On overlong files: makes sense; I'll try to create such a file.
Should I also include the section in `IntervalIntegral` (see #11110)? I have the same problem in that file.

File: Mathlib/MeasureTheory/Integral/SetIntegral.lean [hunk 2]
```diff
@@ -1379,6 +1379,29 @@ variable {X Y F G 𝕜 : Type*} [TopologicalSpace X]
 
 open Metric ContinuousLinearMap
 
+/-- The parametric integral over a continuous function on a compact set is continuous,
+  under mild assumptions on the topologies involved. -/
+theorem continuous_parametric_integral_of_continuous
+    [FirstCountableTopology X] [LocallyCompactSpace X]
+    [OpensMeasurableSpace Y] [SecondCountableTopologyEither Y E] [IsLocallyFiniteMeasure μ]
+    {F : X → Y → E} (hF : Continuous fun p : X × Y ↦ F p.1 p.2) {s : Set Y} (hs : IsCompact s) :
+    Continuous fun x ↦ ∫ a in s, F x a ∂μ := by
```
Comment (loefflerd) [inline Mathlib/MeasureTheory/Integral/SetIntegral.lean:1388]:
```suggestion
    Continuous fun x ↦ ∫ y in s, F x y ∂μ := by
```
or maybe even
```suggestion
    Continuous (∫ y in s, F · y ∂μ) := by

File: Mathlib/MeasureTheory/Integral/SetIntegral.lean [hunk 3]
```diff
@@ -1379,6 +1379,29 @@ variable {X Y F G 𝕜 : Type*} [TopologicalSpace X]
...
+  apply continuousAt_of_dominated
+  · exact eventually_of_forall fun x ↦ (hF.comp (Continuous.Prod.mk x)).aestronglyMeasurable
+  · refine Eventually.mono U_nhds fun x x_in ↦ ?_
+    rw [ae_restrict_iff]
+    · exact eventually_of_forall fun t t_in ↦ hM (mem_image_of_mem _ <| mk_mem_prod x_in t_in)
+    · exact (isClosed_le (hF.comp <| Continuous.Prod.mk x).norm continuous_const).measurableSet
+  · exact integrableOn_const.mpr (Or.inr hs.measure_lt_top)
+  · apply ae_of_all
+    intro a
+    -- TODO: can fun_prop show this?
```
Comment (loefflerd) [inline Mathlib/MeasureTheory/Integral/SetIntegral.lean:1402]: Well, can it? 😄 If not, since `fun_prop` is under active development right now, it might be worth posting this example on Zulip.

File: Mathlib/MeasureTheory/Integral/SetIntegral.lean [hunk 4]
```diff
@@ -1379,6 +1379,29 @@ variable {X Y F G 𝕜 : Type*} [TopologicalSpace X]
...
+  rcases (U_cpct.prod hs).bddAbove_image hF.norm.continuousOn with ⟨M, hM⟩
+  apply continuousAt_of_dominated
+  · exact eventually_of_forall fun x ↦ (hF.comp (Continuous.Prod.mk x)).aestronglyMeasurable
+  · refine Eventually.mono U_nhds fun x x_in ↦ ?_
+    rw [ae_restrict_iff]
+    · exact eventually_of_forall fun t t_in ↦ hM (mem_image_of_mem _ <| mk_mem_prod x_in t_in)
+    · exact (isClosed_le (hF.comp <| Continuous.Prod.mk x).norm continuous_const).measurableSet
+  · exact integrableOn_const.mpr (Or.inr hs.measure_lt_top)
+  · apply ae_of_all
+    intro a
```
Comment (loefflerd) [inline Mathlib/MeasureTheory/Integral/SetIntegral.lean:1401]: Just wondering: can any of this be streamlined with the `filter_upwards` tactic?

Comment (grunweg) [reply Mathlib/MeasureTheory/Integral/SetIntegral.lean]: Thanks for asking! I managed to use it thrice (in the last commit). I'm not sure if all of these are better or if the "using" lines should rather be separate exact statements. Let me know if you prefer a different style!

**Comment by loefflerd** [comment]
Some more comments:
- Something's fishy with the variable declarations at line 1375: you have `X` and `Y` declared twice (not actually harmful, but messy).
- `F` is a normed space in the variable declarations, but a function in `continuous_parametric_integral_of_continuous`; please use lowercase `f`
- I found some other places one can use `fun_prop` and `filter_upwards` in your proof:
````lean
/-- The parametric integral over a continuous function on a compact set is continuous,
  under mild assumptions on the topologies involved. -/
theorem continuous_parametric_integral_of_continuous
    [FirstCountableTopology X] [LocallyCompactSpace X]
    [OpensMeasurableSpace Y] [SecondCountableTopologyEither Y E] [IsLocallyFiniteMeasure μ]
    {f : X → Y → E} (hf : Continuous f.uncurry) {s : Set Y} (hs : IsCompact s) :
    Continuous (∫ y in s, f · y ∂μ) := by
  rw [continuous_iff_continuousAt]
  intro x₀
  rcases exists_compact_mem_nhds x₀ with ⟨U, U_cpct, U_nhds⟩
  rcases (U_cpct.prod hs).bddAbove_image hf.norm.continuousOn with ⟨M, hM⟩
  apply continuousAt_of_dominated
  · filter_upwards with x using Continuous.aestronglyMeasurable (by fun_prop)
  · filter_upwards [U_nhds] with x x_in
    rw [ae_restrict_iff]
    · filter_upwards with t t_in using hM (mem_image_of_mem _ <| mk_mem_prod x_in t_in)
    · exact (isClosed_le (by fun_prop) (by fun_prop)).measurableSet
  · exact integrableOn_const.mpr (Or.inr hs.measure_lt_top)
  · filter_upwards using (by fun_prop)
````
(Interestingly, if one deletes "with x" at the first `filter_upwards` call, the Lean server crashes!)
- Not related to this patch, but close by: is it deliberate that the statement of `continuousOn_integral_bilinear_of_locally_integrable_of_compact_support` requires both `[NormedSpace 𝕜 E]` and `[NormedSpace ℝ E]` at the same time, without requiring any kind of compatibility between `𝕜` and `ℝ`? This feels strange to me. It looks like you never actually use the fact that there's a `𝕜`-linear structure; you just need to know that `L` is Lipschitz in both variables. (But maybe we don't have the API to state that in its natural degree of generality.)

**Comment by grunweg** [comment]
Thank you for the careful observations! Those are nice golfs, I've learned quite a bit :-)

Your last question is a good one. Indeed, the lemma currently uses both the \R-linear and the \k-linear structure.
It is possible to phrase everything over \R; I've pushed that as a temporary commit for you to see. To be honest, I'm not sure what the original intention was: perhaps the author wanted to generalise to others fields where easily possible?

**Comment by loefflerd** [comment]
I just realised that I had misunderstood: I thought `continuousOn_integral_bilinear_of_locally_integrable_of_compact_support` had been added by you, but I see now this was incorrect, it was Sebastian Gouëzel in #8247. So it is clearly irrelevant to the present PR.

**Comment by loefflerd** [comment]
I think brutally enforcing `𝕜 = ℝ` is probably a very bad idea, because the case of complex coefficients is probably needed too. Let's just ignore it and not go looking for trouble 😄

**Comment by loefflerd** [comment]
This looks good to me. 

(Note to maintainers: the issue discussed above about two `NormedSpace` declarations in continuousOn_integral_bilinear_of_locally_integrable_of_compact_support has been queried with Sebastian on zulip, and he says it's intentional; but in any case it is orthogonal to the changes made in this PR.)

maintainer merge

**Comment by grunweg** [comment]
Thank you for the thorough review!

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge
